compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp

```diff
@@ -243,11 +264,106 @@ void DecomposePackUnPackOpsPass::runOnOperation() {
+    return true;
+  }
+  return false;
+}
+
+/// Control function for decomposing pack and unpack ops. Returns true if the
+/// op is a pack or unpack op, and its reshapes can be folded with a producer
+/// or consumer interface tensor op. To be foldable, the following conditions
+/// must be met:
+///
+/// 1. The PackOp or UnPackOp must have no padding.
+/// 2. If the op is a PackOp, then its producer must be a dispatch tensor load.
+/// 3. If the op is an UnPackOp, then all of its consumers must be dispatch
+///    tensor stores.
+/// 4. Any dispatch tensor load producers or dispatch tensor store consumers
+///    must be full slices.
+static LogicalResult isFoldableIntoInterfaceTensor(Operation *op) {
+  // Full slice means zero offsets, unit strides, and sizes match full tensor
+  // shape.
+  auto isFullSlice =
+      [](ArrayRef<OpFoldResult> offsets, ArrayRef<OpFoldResult> sizes,
+         ArrayRef<OpFoldResult> strides, ArrayRef<int64_t> fullTensorShape) {
+        return areAllConstantIntValue(offsets, 0) &&
+               areAllConstantIntValue(strides, 1) &&
+               areConstantIntValues(sizes, fullTensorShape);
+      };
+  if (!isa<tensor::PackOp>(op) && !isa<tensor::UnPackOp>(op)) {
+    return failure();
+  }
+  if (hasPadding(op)) {
+    return failure();
+  }
+
+  // If the producer is a full slice dispatch tensor load, then the `op` is
+  // foldable if it is a PackOp.
+  auto load = dyn_cast<IREE::Flow::DispatchTensorLoadOp>(
+      op->getOperand(0).getDefiningOp());
+  if (isa<tensor::PackOp>(op) && load &&
+      isFullSlice(load.getMixedOffsets(), load.getMixedSizes(),
+                  load.getMixedStrides(), load.getSourceType().getShape())) {
+    return success();
+  }
+  // If all consumers are full slice dispatch tensor stores, then the `op` is
+  // foldable if it is an UnPackOp.
+  if (isa<tensor::UnPackOp>(op) &&
+      llvm::all_of(op->getUsers(), [&](Operation *user) {
+        auto store = dyn_cast<IREE::Flow::DispatchTensorStoreOp>(user);
+        return store &&
+               isFullSlice(store.getMixedOffsets(), store.getMixedSizes(),
+                           store.getMixedStrides(),
+                           store.getTargetType().getShape());
+      })) {
+    return success();
+  }
+  return failure();
+}
+
+void DecomposeBoundaryPackUnPackOpsPass::runOnOperation() {
+  auto funcOp = getOperation();
+  std::optional<OpPassManager> maybeFuncPassManager =
+      getFunctionOpInterfacePassManager(funcOp);
+  if (!maybeFuncPassManager) {
+    funcOp.emitOpError("unhandled operation type while creating pass pipeline "
+                       "nested on `FunctionOpInterface`");
+    return signalPassFailure();
+  }
+  OpPassManager &funcPassManager = maybeFuncPassManager.value();
+  funcPassManager.addPass(createDecomposePackUnPackOpsPass(
+      tileOuterToOne, /*useOnlyReshapes=*/true,
+      /*controlFn=*/isFoldableIntoInterfaceTensor));
```

Comment:
Why do you need a whole pass for this? You're just using the control function on the existing pass; I would factor out this pass and move the control function into a shared location.

Reply:
The reason to have a separate pass is so that it is possible to run the pass using the command line. The other way to do this would be to add an option pointing to a specific control function, but I opted for adding a new pass because `useOnlyReshapes` does not exist in DecomposeBoundaryPackUnPackOps (since it should always be true for this new pass), so I thought it was more clear this way.

Reply:
Invoking a pass inside another pass is not the best way to do that then because it leads to confusing IR dumps. Factor DecomposePackUnPackOpsPass::runOnOperation into its own function and call the same function from this pass with different options. The other way to do it is with custom test pass pipelines, but the logic in the control function is involved enough that a separate pass makes sense

Reply:
I created a shared runOnOperation function, so now there is no nested pass call. PTAL